tianshou/policy/modelfree/bcq.py

```diff
@@ -0,0 +1,138 @@
+import torch
+import numpy as np
+import torch.nn.functional as F
+from typing import Any, Dict, Union, Optional
+
+from tianshou.data import Batch, to_torch
+from tianshou.policy import BasePolicy
+import torch.nn as nn
+from copy import deepcopy
+
+
+class BCQPolicy(BasePolicy):
+    """Implementation discrete BCQ algorithm. Some code is from
+    https://github.com/sfujim/BCQ/tree/master/discrete_BCQ
+
+    """
+
+    class BCQN(nn.Module):
+        """BCQ NN for dialogue policy. It includes a net for imitation and a net for Q-value"""
+
+        def __init__(
+            self, input_size, n_actions, imitation_model_hidden_dim, policy_model_hidden_dim
+        ):
+            super(BCQPolicy.BCQN, self).__init__()
+            self.q1 = nn.Linear(input_size, policy_model_hidden_dim)
+            self.q2 = nn.Linear(policy_model_hidden_dim, policy_model_hidden_dim)
+            self.q3 = nn.Linear(policy_model_hidden_dim, n_actions)
+
+            self.i1 = nn.Linear(input_size, imitation_model_hidden_dim)
+            self.i2 = nn.Linear(imitation_model_hidden_dim, imitation_model_hidden_dim)
+            self.i3 = nn.Linear(imitation_model_hidden_dim, n_actions)
+
+        def forward(self, state):
+            q = F.relu(self.q1(state))
+            q = F.relu(self.q2(q))
+
+            i = F.relu(self.i1(state))
+            i = F.relu(self.i2(i))
+            i = F.relu(self.i3(i))
+            return self.q3(q), F.log_softmax(i, dim=1), i
```

Comment:
What do you mean by moving to `test/`? I think in other modules, taking dqn for example, it uses the pytorch model defined in `tianshou/utils/net/common.py`. Are you saying I should move the pytorch model definition here to `tianshou/utils/net/common.py` as well? The reason I define it here is that the pytorch model used in BCQ is different from others -- it actually includes two nets instead of one. So it's pretty much BCQ specific and not common at all.

Reply:
I think you can take a look at the current code of TD3/SAC. They both use Q-learning and double critic network.